### ISSUE TITLE
Support borrowing within internally tagged enum

### ISSUE THREAD

**Issue by jmi2k** [serde-rs/serde#906]
Serde panics when trying to deserialize an struct which has a borrowed field, which also contains a borrowed field. For example, the following code compiles fine:

**EDIT:** fixed the example
```rust
#[macro_use]
extern crate serde_derive;

extern crate serde;
extern crate serde_json;

#[derive(Deserialize, Debug)]
#[serde(tag = "type")]
pub enum Input<'a> {
    Package { name: &'a str },
}

fn main() {
    let j = r#"{"type":"Package","name":""}"#;
    println!("{}", serde_json::from_str::<Input>(j).unwrap_err());
}
```

But you get the following panic:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { inner: ErrorInner { kind: Custom, line: None, col: 0, message: "invalid type: string \"bye\", expected a borrowed string", key: ["input"] } }', /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libcore/result.rs:868
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

**Comment by dtolnay** [comment]
I don't think this is correctly minimized from what you showed in IRC earlier. [It gives a different error.](http://play.integer32.com/?gist=29b4a709bd0aecba7a5aa19024619778) Side note: the panic is coming from your code not Serde, because you unwrap the error. :wink:

**Comment by dtolnay** [comment]
I minimized it to this:

```rust
#[macro_use]
extern crate serde_derive;

extern crate serde;
extern crate serde_json;

#[derive(Deserialize, Debug)]
#[serde(tag = "type")]
pub enum Input<'a> {
    Package { name: &'a str },
}

fn main() {
    let j = r#"{"type":"Package","name":""}"#;
    println!("{}", serde_json::from_str::<Input>(j).unwrap_err());
}
```

```
invalid type: string "", expected a borrowed string
```

**Comment by jmi2k** [comment]
Oh sorry, the code in the link is indeed wrong. I think it has saved changes I made after pasting it in IRC...
And yes, that's _real_ minimization what you have done.

**Comment by dtolnay** [comment]
@jmi2k I released the fix in Serde 1.0.4 and confirmed that the minimized example now works.

**Comment by jmi2k** [comment]
Thanks @dtolnay :D
